File: atcoder.jp/contests/abc176/tasks/abc176_c/main.ts
import { readFileSync } from "fs";
let inputLinesIndex = 0;
const inputLines = readFileSync("/dev/stdin", "utf8").split("\n");
const input = () => inputLines[inputLinesIndex++];

const scanl1 = <T>(func: (a: T, b: T) => T, array: T[]) => {
  const ret = [array[0]];
  array.reduce((a, b) => {
    const x = func(a, b);
    ret.push(x);
    return x;
  });
  return ret;
};
const sum = (array: number[]) => array.reduce((a, b) => a + b, 0);
const zip: {
  (a: string, b: string): [string, string][];
  <B>(a: string, b: B[]): [string, B][];
  <A>(a: A[], b: string): [A, string][];
  <A, B>(a: A[], b: B[]): [A, B][];
} = <A, B>(a: string | A[], b: string | B[]) =>
  typeof a === "string" && typeof b === "string"
    ? [...Array(Math.min(a.length, b.length))].map(
        (_, i) => [a[i], b[i]] as [string, string]
      )
    : typeof a === "string" && typeof b !== "string"
    ? [...Array(Math.min(a.length, b.length))].map(
        (_, i) => [a[i], b[i]] as [string, B]
      )
    : typeof a !== "string" && typeof b === "string"
    ? [...Array(Math.min(a.length, b.length))].map(
        (_, i) => [a[i], b[i]] as [A, string]
      )
    : [...Array(Math.min(a.length, b.length))].map(
        (_, i) => [a[i], b[i]] as [A, B]
      );

const main = () => {
  const N = +input();
  const A = input().split(" ").map(Number);

  const accMax = scanl1(Math.max, A);
  const ans = sum(zip(accMax, A).map(([a, b]) => a - b));

  console.log(ans);
};

main();

File: atcoder.jp/contests/abc176/tasks/abc176_c/main.cpp
#include <bits/stdc++.h>
using namespace std;

using ll = long long;

template <typename T>
vector<T> scanl1(function<T(T, T)> func, const vector<T> &vec) {
  vector<T> result(vec.size());
  partial_sum(vec.begin(), vec.end(), result.begin(), func);
  return result;
}

template <typename T1, typename T2>
vector<pair<T1, T2>> zip(const vector<T1> &a, const vector<T2> &b) {
  size_t n = min(a.size(), b.size());
  vector<pair<T1, T2>> ret(n);
  for (size_t i = 0; i < n; i++) ret[i] = {a[i], b[i]};
  return ret;
}

vector<int> input_vector(int n) {
  vector<int> ret(n);
  for (auto &&i : ret) cin >> i;
  return ret;
}

int main() {
  int N;
  cin >> N;
  auto A = input_vector(N);

  auto acc_max = scanl1<int>([](int a, int b) { return max(a, b); }, A);

  ll ans = 0;
  for (auto &&[a, b] : zip(acc_max, A)) ans += a - b;

  cout << ans << endl;
}
